### PR TITLE
Bill Mailer

### LINKED ISSUE #8: Send Bills to recipient via email

### PR DESCRIPTION
**Pull request by soberstadt**
Closes #8 and #17 🥳

### CODE REVIEW

**Comment by soberstadt** [comment]
Tested and working great :D 
image

**Review by dbenton9 — COMMENTED** [review]
I'm confused, in your commits it appears that you've added the functionality to generate multiple bills and email them but when I clone and look at this locally I don't see the new features... but maybe you've not made them accessible because it doesn't appear this PR has adjusted any of the user views. What's your plan there?

Screen Shot 2021-07-12 at 12 20 36 PM

**Comment by soberstadt** [comment]
My plan was to get this much merged and then to make another PR that actually exposes it.

File: spec/mailers/previews/bill_mailer_preview.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+  sig { returns(ActionMailer::MessageDelivery) }
+  def bill
+    BillMailer.bill(
+      to_name: 'Bill',
+      to_email: 'bill@cru.org',
+      from: 'Vonette',
+      for_message: 'Helping me learn to backflip!',
+      count: 7
+    )
```
Comment (dbenton9) [inline spec/mailers/previews/bill_mailer_preview.rb:14]: will this be replaced by user input?

File: config/environments/production.rb [hunk 2]
```diff
@@ -56,6 +56,20 @@
...
+
+  # SMTP settings for gmail
+  config.action_mailer.smtp_settings = {
+    address: 'smtp.gmail.com',
+    port: 587,
+    user_name: Rails.application.credentials.google[:email],
+    password: Rails.application.credentials.google[:password],
+    authentication: 'plain',
+    enable_starttls_auto: true
+  }
```
Comment (dbenton9) [inline config/environments/production.rb:71]: I've never seen a mailer setup before. To send emails do we have to have an email address for the sender?

File: app/views/bill_mailer/bill.html.erb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<p>Yay! Someone sent you some Bill Bucks!!!</p>
+
+<% @bucks.keys.each do |filename| %>
+<%= image_tag(attachments[filename].url, size: "300x288") %>
+<% end %>
```
Comment (dbenton9) [inline app/views/bill_mailer/bill.html.erb:5]: Is this the email content people receive?

File: app/mailers/bill_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      hash[vonette.filename(i)] = vonette
+      i += 1
+    end
+    ones.times do
+      hash[bill.filename(i)] = bill
+      i += 1
+    end
+
+    hash
+  end
```
Comment (dbenton9) [inline app/mailers/bill_mailer.rb:56]: So based on the `count` provided this selects the optimal vonette(s)/bill(s) combo and generate those images?

File: app/mailers/bill_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    mail(to: to_email, subject: "You've got Bills (Bucks)")
+  end
+
+  private
+
+  sig do
+    params(to_name: String, from: String, for_message: String, count: Integer)
+      .returns(T::Hash[String, Buck])
+  end
+  def bucks(to_name:, from:, for_message:, count: 1)
```
Comment (dbenton9) [inline app/mailers/bill_mailer.rb:36]: Why do this in the mailer? I thought we already had the model/controller creating bucks?

**Review by dbenton9 — COMMENTED** [review]
First time looking at mailer stuff so I've got some newbie questions. From the looks of this you can only generate multiple bucks if you're using the mailer feature, but for just using the regular BB generator we'll still be confined to generating one type of buck at a time. Would we eventually update the current BB generator to support multiple BB creations?

File: spec/mailers/previews/bill_mailer_preview.rb (hunk 1 above)
Comment (soberstadt) [reply spec/mailers/previews/bill_mailer_preview.rb:14]: The preview file is a cool thing that rails has built in for previewing Mailers. You define a call to a mailer action in this class and rails generates this for you: http://localhost:3001/rails/mailers/bill_mailer/bill.html

File: config/environments/production.rb (hunk 2 above)
Comment (soberstadt) [reply config/environments/production.rb:71]: This is the configuration for Rails to actually be able to send an email. I'm using a "service"
google account: bill.buck.generator@gmail.com

File: app/views/bill_mailer/bill.html.erb (hunk 3 above)
Comment (soberstadt) [reply app/views/bill_mailer/bill.html.erb:5]: Yes.

File: app/mailers/bill_mailer.rb (hunk 5 above)
Comment (soberstadt) [reply app/mailers/bill_mailer.rb:36]: Good point, I thought about making a BuckCollection class that can be passed from the controller to the mailer. I'll give that a try.

**Pull request merged**